### PR TITLE
Add `ignore_idx` to retrieval metrics

### LINKED ISSUE #672: Add `ignore_idx` to retrieval metrics
## 🚀 Feature

Add `ignore_idx` as keyword argument to all `retrieval` metrics. Instantiation will be like `RetrievalMAP(ignore_idx=-100)`. Entries where the labels are equal to `ignore_idx` will not be considered.

### Motivation

It is common to have some examples where the predictions should not be considered.

### Pitch

Predictions where the labels are equal to `ignore_idx` will be skipped.

### Additional context

Many other metrics include this feature.

### PR DESCRIPTION
**Pull request by lucadiliello**
## What does this PR do?

Fixes #672 

## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure to update the docs?
- [X] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Comment by lucadiliello** [comment]
Seems to me that we should reduce the number of tests, or maybe reduce their complexity. Do you know whether I can check how much time was taken just by the `retrieval` package?

**Comment by Borda** [comment]
> Seems to me that we should reduce the number of tests, or maybe reduce their complexity. Do you know whether I can check how much time was taken just by the `retrieval` package?

yes, it becomes a heavy bit I guess the most is coming from Image and Text packages...
you can run pytest with ` --durations=0` which ends with times over all/each test and you can aggregate it... just thinking about it, it would be a great analysis for us to identify what shall we improve :rabbit: 

on the other note, I am not very sure if we shall lower test coverage, rather we can think about running some job concurrency and reducing configuration with a limited number of agents (the MacOs is limited and also the longest)

File: torchmetrics/retrieval/retrieval_hit_rate.py [hunk 1]
```diff
@@ -75,6 +77,7 @@ class RetrievalHitRate(RetrievalMetric):
     def __init__(
         self,
         empty_target_action: str = "neg",
+        ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_hit_rate.py:80]:
```suggestion
        ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_metric.py [hunk 2]
```diff
@@ -71,6 +73,7 @@ class RetrievalMetric(Metric, ABC):
     def __init__(
         self,
         empty_target_action: str = "neg",
+        ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_metric.py:76]:
```suggestion
        ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_ndcg.py [hunk 3]
```diff
@@ -75,6 +77,7 @@ class RetrievalNormalizedDCG(RetrievalMetric):
     def __init__(
         self,
         empty_target_action: str = "neg",
+        ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_ndcg.py:80]:
```suggestion
        ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_precision.py [hunk 4]
```diff
@@ -75,6 +77,7 @@ class RetrievalPrecision(RetrievalMetric):
     def __init__(
         self,
         empty_target_action: str = "neg",
+        ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_precision.py:80]:
```suggestion
        ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_recall.py [hunk 5]
```diff
@@ -75,6 +77,7 @@ class RetrievalRecall(RetrievalMetric):
     def __init__(
         self,
         empty_target_action: str = "neg",
+        ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_recall.py:80]:
```suggestion
        ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_recall.py [hunk 6]
```diff
@@ -75,6 +77,7 @@ class RetrievalRecall(RetrievalMetric):
...
         k: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_recall.py:81]:
```suggestion
        k: Optional[int] = None,
```

File: torchmetrics/utilities/checks.py [hunk 7]
```diff
@@ -516,13 +516,15 @@ def _check_retrieval_inputs(
     preds: Tensor,
     target: Tensor,
     allow_non_binary_target: bool = False,
+    ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/utilities/checks.py:519]:
```suggestion
    ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_precision.py [hunk 8]
```diff
@@ -75,6 +77,7 @@ class RetrievalPrecision(RetrievalMetric):
...
         k: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_precision.py:81]:
```suggestion
        k: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_ndcg.py [hunk 9]
```diff
@@ -75,6 +77,7 @@ class RetrievalNormalizedDCG(RetrievalMetric):
...
         k: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_ndcg.py:81]:
```suggestion
        k: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_hit_rate.py [hunk 10]
```diff
@@ -75,6 +77,7 @@ class RetrievalHitRate(RetrievalMetric):
...
         k: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_hit_rate.py:81]:
```suggestion
        k: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_fallout.py [hunk 11]
```diff
@@ -77,6 +79,7 @@ class RetrievalFallOut(RetrievalMetric):
     def __init__(
         self,
         empty_target_action: str = "pos",
+        ignore_index: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_fallout.py:82]:
```suggestion
        ignore_index: Optional[int] = None,
```

File: torchmetrics/retrieval/retrieval_fallout.py [hunk 12]
```diff
@@ -77,6 +79,7 @@ class RetrievalFallOut(RetrievalMetric):
...
         k: int = None,
```
Comment (SkafteNicki) [inline torchmetrics/retrieval/retrieval_fallout.py:83]:
```suggestion
        k: Optional[int] = None,
```

**Review by SkafteNicki — COMMENTED** [review]
Each metric have a `Raises` section in their docstring, where value error for `ignore_index` should be added.

**Review by SkafteNicki — APPROVED** [review]

**Pull request merged**